bus-mapping/src/lib.rs

```diff
@@ -0,0 +1,177 @@
+//! Crate docs
+
+#![cfg_attr(docsrs, feature(doc_cfg))]
+// Temporary until we have more of the crate implemented.
+#![allow(dead_code)]
+// Catch documentation errors caused by code changes.
+#![deny(broken_intra_doc_links)]
+#![deny(missing_debug_implementations)]
+#![deny(missing_docs)]
+#![deny(unsafe_code)]
+
+use std::{
+    collections::BTreeMap,
+    iter::Iterator,
+    ops::{Index, IndexMut},
+};
+
+use itertools::Itertools;
+use pasta_curves::arithmetic::FieldExt;
+
+// -------- EVM Circuit
+// gc as index of each row corresponds to the state after the opcode is performed
+// 1 gc has 1 entry
+//`StackElem{key 	val 	stack_op(READ/WRITE) 	gc   OPCODE    CallID(later_future)}`
+
+//------ State Circuit
+// Group target (stack or memory)
+// Group by key (location in memory/storage)
+// Sorty by gc
+//`MemoryElem{target 	gc 	val1 	val2 	val3}`
+
+#[derive(Debug, Clone)]
+struct BlockConstants<F: FieldExt> {
+    hash: [u8; 256], // Until we know how to deal with it
+    coinbase: F,
+    timestamp: F,
+    number: F,
+    difficulty: F,
+    gaslimit: F,
+    chain_id: F,
+}
+
+#[derive(Debug, Clone)]
+enum RW {
+    READ,
+    WRITE,
+}
+
+/// Doc
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub enum Target {
+    /// Doc
+    Memory,
+    /// Doc
+    Stack,
+    /// Doc
+    Storage,
+}
+
+/// Doc
+#[derive(Debug, Clone)]
+pub struct Operation<F: FieldExt> {
+    mem_op: RW,
+    target: Target,
+    key: usize,
+    value: F,
+    opcode_info: String,
+}
+
+/// Bus Mapping structure
+#[derive(Debug, Clone)]
+pub struct BusMapping<F: FieldExt> {
+    entries: Vec<Operation<F>>,
+    block_ctants: BlockConstants<F>,
+    // Helper to sort by key groups. We store how many different keys we have when building the Bus Mapping
+    mem_ops_sorted: BTreeMap<usize, Vec<Operation<F>>>,
+    stack_ops_sorted: BTreeMap<usize, Vec<Operation<F>>>,
+    storage_ops_sorted: BTreeMap<usize, Vec<Operation<F>>>,
+}
+
+impl<F: FieldExt> Index<usize> for BusMapping<F> {
+    type Output = Operation<F>;
+    fn index(&self, index: usize) -> &Self::Output {
+        &self.entries[index]
+    }
+}
+
+impl<F: FieldExt> IndexMut<usize> for BusMapping<F> {
+    fn index_mut(&mut self, index: usize) -> &mut Self::Output {
+        &mut self.entries[index]
+    }
+}
+
+impl<F: FieldExt> From<(Vec<Operation<F>>, BlockConstants<F>)> for BusMapping<F> {
+    fn from(inp: (Vec<Operation<F>>, BlockConstants<F>)) -> Self {
+        // Initialize the BTreeMaps with empty vecs for each key group
+        let mut mem_ops_sorted = BTreeMap::new();
+        let mut stack_ops_sorted = BTreeMap::new();
+        let mut storage_ops_sorted = BTreeMap::new();
+        inp.0
+            .iter()
+            .map(|op| op.key)
+            .unique()
+            .sorted()
+            .for_each(|key| {
+                mem_ops_sorted.insert(key, vec![]);
+                stack_ops_sorted.insert(key, vec![]);
+                storage_ops_sorted.insert(key, vec![]);
```

Comment:
In this way, we init an empty vec for `key` even the operation is not for the `target`, so we need to filter out empty vec for each `{mem,stack,storage}_ops_sorted` when calling `{mem,stack,storage}_part`

Maybe we could just sort them by `(target, key)` and group them by matching like this:
```rust
inp.0
    .clone()
    .into_iter()
    .sorted_by_key(|op| (op.target, op.key))
    .for_each(|op| match op.target {
        Target::Memory => {
            mem_ops_sorted.entry(op.key).or_default().push(op);
        }
        Target::Stack => {
            stack_ops_sorted.entry(op.key).or_default().push(op);
        }
        Target::Storage => {
            storage_ops_sorted.entry(op.key).or_default().push(op);
        }
    });
```
Then we don't need the loops below.

Reply:
Also should we push `index` into `{mem,stack,storage}_op_sorted` as `gc`? Otherwise we don't know what `gc` is when calling `{mem,stack,storage}_part`

Reply:
Ohhh how did I miss that!! That's completely right! Changing it immediately! 

> Also should we push index into {mem,stack,storage}_op_sorted as gc? Otherwise we don't know what gc is when calling {mem,stack,storage}_part

Here is my dilema:

Idk wether I should create a new struct that contains the gc or force the `Operation` to contain it. I didn't want it to have it inside since I wanted to have a single source of truth about the gc. And if We have it inside the `Operation` we have two, the field inside the struct and the index we iterate over. 

Any ideas on which would be the best choice?

Reply:
Totally understand. If we assume bus mapping would only be pushed with more elements instead of mutating existing ones, could we just have `{mem,stack,storage}_op_sorted` store `Vec<(usize /* gc */, Operation)>`?